dali/imgcodec/util/convert_gpu_test.cc

```diff
@@ -0,0 +1,303 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+#include <string>
+#include <vector>
+#include <memory>
+#include "dali/test/dali_test.h"
+#include "dali/test/dali_test_config.h"
+#include "dali/core/tensor_shape_print.h"
+#include "dali/core/convert.h"
+#include "dali/imgcodec/util/convert_gpu.h"
+#include "dali/test/test_tensors.h"
+#include "dali/test/tensor_test_utils.h"
+#include "dali/core/cuda_stream_pool.h"
+#include "dali/kernels/imgproc/color_manipulation/color_space_conversion_impl.h"
+
+namespace dali {
+namespace imgcodec {
+namespace test {
+
+namespace {
+
+template<class Input, class Output>
+struct ConversionTestType {
+  using In = Input;
+  using Out = Output;
+};
+
+using TensorTestData = std::vector<std::vector<std::vector<float>>>;
+
+template<class T>
+void init_test_tensor_list(kernels::TestTensorList<T> &list, const TensorTestData &data) {
+  TensorShape<> shape = {
+    static_cast<int>(data.size()),
+    static_cast<int>(data[0].size()),
+    static_cast<int>(data[0][0].size()),
+  };
+  list.reshape({{shape}});
+  auto tv = list.cpu()[0];
+  for (int i = 0; i < shape[0]; i++)
+    for (int j = 0; j < shape[1]; j++)
+      for (int k = 0; k < shape[2]; k++)
+        *tv(TensorShape<>{i, j, k}) = ConvertSatNorm<T>(data[i][j][k]);
+}
+
+template<class T>
+SampleView<GPUBackend> get_gpu_sample_view(kernels::TestTensorList<T> &list) {
+  auto tv = list.gpu()[0];
+  return SampleView<GPUBackend>(tv.data, tv.shape, type2id<T>::value);
+}
+
+}  // namespace
+
+template <typename ConversionType>
+class ConvertGPUTest : public ::testing::Test {
+ public:
+  using Input = typename ConversionType::In;
+  using Output = typename ConversionType::Out;
+
+  void SetReference(const TensorTestData &data) {
+    init_test_tensor_list(reference_list_, data);
+    output_list_.reshape({{reference_list_.cpu()[0].shape}});
+  }
+
+  void SetInput(const TensorTestData &data) {
+    init_test_tensor_list(input_list_, data);
+  }
+
+  void CheckConvert(TensorLayout out_layout, DALIImageType out_format,
+                    TensorLayout in_layout, DALIImageType in_format,
+                    const ROI &roi = {}, float multiplier = 1.0f) {
+    int device_id;
+    CUDA_CALL(cudaGetDevice(&device_id));
+    auto out = get_gpu_sample_view(output_list_);
+    auto in = get_gpu_sample_view(input_list_);
+    auto stream = CUDAStreamPool::instance().Get(device_id);
+    Convert(out, out_layout, out_format, in, in_layout, in_format, stream, roi, multiplier);
+    CUDA_CALL(cudaStreamSynchronize(stream));
+    Check(output_list_.cpu()[0], reference_list_.cpu()[0], EqualConvertNorm(eps_));
```

Comment:
I think you were removing EqualConvertNorm in other places, is it wanted here?

Reply:
I was changing direct Check calls to use the main AssertClose overload, so the behaviour is the same (dividing the epsilon).